pkg/cmd/roachtest/tests/backup.go

```diff
@@ -164,58 +171,351 @@ func registerBackupNodeShutdown(r registry.Registry) {
-				// Backup from old node should succeed.
-				successfulBackupStep(3),
-				// Backup from old node with revision history should succeed.
-				successfulBackupStep(4, "revision_history"),
+				setShortJobIntervalsStep(1),
+				loadBackupDataStep(c),
+				// Upgrade some nodes.
+				binaryUpgradeStep(upgradedNodes, mainVersion),
+
+				// Let us first test planning and executing a backup on different node
+				// versions.
+				//
+				// NB: All backups in this test are writing to node 1's ExternalIODir
+				// for simplicity.
+
+				// Case 1: plan backup    -> old node
+				//         execute backup -> upgraded node
+				//
+				// Halt job execution on older nodes.
+				disableJobAdoptionStep(c, oldNodes),
+
+				// Run a backup from an old node so that it is planned on the old node
+				// but the job is adopted on a new node.
+				planAndRunBackup(t, c, oldNodes.RandNode(), oldNodes,
+					`BACKUP TABLE bank.bank INTO 'nodelocal://1/plan-old-resume-new' WITH detached`),
+
+				// Run an incremental backup from an old node so that it is planned on the old node
+				// but the job is adopted on a new node.
+				planAndRunBackup(t, c, oldNodes.RandNode(), oldNodes,
+					`BACKUP TABLE bank.bank INTO LATEST IN 'nodelocal://1/plan-old-resume-new' WITH detached`),
+
+				enableJobAdoptionStep(c, oldNodes),
+
+				// Case 2: plan backup    -> upgraded node
+				//         execute backup -> old node
+				//
+				// Halt job execution on upgraded nodes.
+				disableJobAdoptionStep(c, upgradedNodes),
+
+				// Run a backup from a new node so that it is planned on the new node
+				// but the job is adopted on an old node.
+				planAndRunBackup(t, c, upgradedNodes.RandNode(), upgradedNodes,
+					`BACKUP TABLE bank.bank INTO 'nodelocal://1/plan-new-resume-old' WITH detached`),
+
+				// Run an incremental backup from a new node so that it is planned on
+				// the new node but the job is adopted on an old node.
+				planAndRunBackup(t, c, upgradedNodes.RandNode(), upgradedNodes,
+					`BACKUP TABLE bank.bank INTO LATEST IN 'nodelocal://1/plan-new-resume-old' WITH detached`),
+
+				enableJobAdoptionStep(c, upgradedNodes),
+
+				// Now let us test building an incremental chain on top of a full backup
+				// created by a node of a different version.
+				//
+				// Case 1: full backup -> new nodes
+				//         inc backup  -> old nodes
+				disableJobAdoptionStep(c, oldNodes),
+				// Plan and run a full backup on the new nodes.
+				planAndRunBackup(t, c, upgradedNodes.RandNode(), oldNodes,
+					`BACKUP TABLE bank.bank INTO 'nodelocal://1/new-node-full-backup' WITH detached`),
+				// Set up the cluster so that only the old nodes plan and run the
+				// incremental backup.
+				enableJobAdoptionStep(c, oldNodes),
+				disableJobAdoptionStep(c, upgradedNodes),
+
+				// Run an incremental (on old nodes) on top of a full backup taken by
+				// nodes on the upgraded version.
+				planAndRunBackup(t, c, oldNodes.RandNode(), upgradedNodes,
```

Comment:
If the table keeps changing fingerprinting at the end becomes a little tricky because we have to then store the fingerprint at the time the backup was taken, to then compare it against the restored table. I can check if that's simple enough to do.

Reply:
sweet. This suggestion also applies to the cross-version incremental backup cases further below.

Reply:
👍 I have something that's not great but should work.

Reply:
Turned out to be a lot harder than I thought it would be because of the way the `newVersionUpgradeTest` captures each version upgrade function, and in turn their parameters. I had to pass around pointers to the `fingerprints` slice to ensure that the verify backup step  at the end would see the fingerprints we aggregated through the course of the test.

Reply:
changed to a map based on your offline suggestion, tftr!